ocaml/xapi/cert_distrib.ml

```diff
@@ -399,29 +399,115 @@ let exchange_certificates_among_all_members ~__context =
     (fun host -> Worker.remote_regen_bundle host rpc session_id)
     all_hosts
 
-let import_joiner ~__context ~uuid ~certificate ~to_hosts =
-  let joiner_certificate =
-    HostPoolProvider.certificate_of_id_content uuid certificate
+let ( (get_local_ca_certs : unit -> WireProtocol.certificate_file list)
+    , (get_local_pool_certs : unit -> WireProtocol.certificate_file list) ) =
+  let g path () =
+    (* collects all certs in [path] ending in "pem". this is equivalent to the
+     * certs that would be put in a bundle, if update-ca-bundle.sh were to be
+     * executed on [path] *)
+    Sys.readdir path
+    |> Array.to_list
+    |> List.filter (Fun.flip Filename.check_suffix "pem")
+    |> List.map (fun filename ->
+           let path = Filename.concat path filename in
+           let content = string_of_file path in
+           WireProtocol.{filename; content}
+       )
+  in
+  (g ApplianceProvider.store_path, g HostPoolProvider.store_path)
+
+let am_i_missing_certs ~__context : bool =
+  (* compare what's in the database with what's on my filesystem *)
+  let missing_pool_certs () =
+    let expected_files =
+      Db.Host.get_all ~__context
+      |> List.map (fun self -> Db.Host.get_uuid ~__context ~self)
+      |> List.map (fun x -> x ^ ".pem")
+      |> StringSet.of_list
+    in
+    let actual_files =
+      Sys.readdir HostPoolProvider.store_path
+      |> Array.to_seq
```

Comment:
Don't we need to filter based on extension here, too?

Reply:
I don't think it matters whether or not we remove the 'non .pems' from this set: the strings in `expected_files` all end in ".pem", so `expected_files` is a subset of `actual_files` if and only if it is a subset of `Set.union actual_files S`, where `S` is any set containing strings not ending in `.pem`

(same applies below)